Solr-based suggestion / autocomplete

kshepherd opened the pull request:
## Solr suggest handler and discovery method

* Requires https://github.com/DSpace/dspace-angular/pull/4439
* Rest Contract doc: https://github.com/DSpace/RestContract/pull/328
* Related to #10237 
* (they solve similar issues in slightly different ways)

This pull request adds the ability to use the Solr suggest request handler to generate search suggestions, input suggestions using the `HighFrequencyDictionaryFactory` for weighted (by freq) terms from the search Solr core, or `FileDictionaryFactory` for text-file based vocabularies.

This allows for flexible use in a few different ways, and currently is added into Submission / Edit Item forms for input field suggestions using typeahead, using the following approach:

**Vocabulary**
* The vocabulary model and services are extended to provide a 'type' so "suggest" can be treated differently without affecting other usage

**Solr and discovery**
* The suggest handler is enabled for Solr search, with some `_suggest` fields - each suggest "dictionary" config needs to be added to `solrconfig.xml` with field name and other options set
* Configuration tells a Solr index plugin which fields to copy into `_suggest` dynamic Solr fields

**REST**
* A new /suggest method on the DiscoveryRestRepository is added to take a query and dictionary, and will return plain JSON with suggestions - the suggestions are not strongly modelled as this is intended to be a light-weight solution that does not involve addressable objects
* (new corresponding functions are added to the angular search service)

**Submission**
The submission-form XML is extended to take a "vocabulary type" when declaring a field to be controlled by a vocabulary:
```
 <!ATTLIST vocabulary
     closed (true|false) "false"
     type (xml|authority|suggest) "xml">
```
* XML should indicate the plain, legacy XML files (maybe obsolete since I think most just use the ChoiceAuthority plugin now anyway?
* Authority indicates to use the authority plugin (this could allow for #10237 to be referenced for facet autocomplete, or the existing tree lookups, etc.)
* Suggest indicates that the Discovery suggest method should be used, with the vocabulary name as the dictionary

There is also a new `validation-dictionary` element which can reference a suggest dictionary to validate saved values against.

**UI**
In the `dynamic-onebox` Angular component which traditionally handles vocabulary / authority display like tree lookup or autocomplete, if the vocabulary type is "suggest", it will use a different search method and display template for the suggest terms. If not, the current behaviour is used (lookup / suggest from XML / etc)

![f31100b4-99df-4f9e-91cb-341cb6b66c3b](https://github.com/user-attachments/assets/82d4cb4e-d8c5-4a23-a679-4b94b1dbcb03)

The fuzzy search can match multiple terms, highlight the matching term parts and sort the matches by frequency which is useful for suggesting the most common metadata values first.


Developed by The Library Code with the support of Universität Bremen and GESIS.

## How to Test

* See the top of this issue for the required frontend PR
* It is recommended to use the solrconfig.xml and schema.xml as supplied in the search core in this PR. Depending on how you deploy/use solr, you may need to recreate the search core and fully reindex, to ensure the new schema is active
* In order to test dc.subject as a metadata dictionary, you should have a decent number of values (unique and non-unique) in dc.subject fields
* `dc.subject` is enabled for indexing to `dc.subject_suggest` by default, you can check this in `discovery.cfg`
* The example `<vocabulary type="suggest"....` usage is available in a comment in `traditionalpagetwo` of `submission-forms.xml` - for the quickest test, simply uncomment this line and try out submissions of a standard item that uses the traditionalpagetwo submission step / form.
* You can also try out the "countries_file" vocabulary name instead of "subject" to see how the feature works by indexing a flat txt file instead of metadata values from the search core

## Open questions

* Should we leave discovery.suggest.fields and discovery.suggest.allowed-dictionaries blank by default with commented examples? If so we would need to override the test dspaceFolder with another discovery.cfg that enabled them, or modify our test suite to programmatically add/set those values. Right now they make reviewing a bit easier so I've left them as is.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

kshepherd committed:
Autocomplete suggestions from Solr

# Conflicts:
#	dspace-api/src/main/java/org/dspace/content/authority/ChoiceAuthorityServiceImpl.java
#	dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/SubmissionFormConverter.java
#	dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/AuthorityUtils.java
#	dspace/config/modules/authority.cfg
#	dspace/config/modules/discovery.cfg
#	dspace/config/spring/api/discovery.xml
#	dspace/solr/search/conf/schema.xml

kshepherd committed:
Autocomplete solr suggestion ITs

kshepherd committed:
Autocomplete solr suggestion flat file example

milanmajchrak committed:
fix: add security and robustness to suggest endpoint

1. Authentication: require login via @PreAuthorize (prevents anonymous abuse)
2. Dictionary allowlist: configurable via discovery.suggest.allowed-dictionaries
   - Non-allowed dictionaries return HTTP 403
3. Input validation: blank/missing query or dict params return HTTP 400
4. Deduplication: remove duplicate suggestions from Solr response
5. Error handling: catch Solr exceptions, return HTTP 500 instead of stacktrace
6. Integration tests for auth (401), allowlist (403), validation (400), success (200)

EikLoe commented:
Hi @kshepherd ,
thanks for your PR, it tested it locally the subject suggestion did work for me. I tested it using the Frontend and via curl on the rest-api.
However when I tried to use the _countries.txt_ file (http://localhost:8080/server/api/discover/suggest?dict=countries_file&q=Fra). I don't get any results and I noticed the following error in the solr logs while starting the server:
```
2026-02-27 11:37:05.137 ERROR (searcherExecutor-19-thread-1-processing-search) [c: s: r: x:search t:] o.a.s.h.c.SuggestComponent Exception in building suggester index for countries_file:  => java.lang.RuntimeException: org.apache.solr.core.SolrResourceNotFoundException: Can't find resource 'countries.txt' in classpath or '/var/solr/data/search'
2026-02-27T11:37:05.156518967Z 	at org.apache.solr.spelling.suggest.FileDictionaryFactory.create(FileDictionaryFactory.java:58)
2026-02-27T11:37:05.156529758Z java.lang.RuntimeException: org.apache.solr.core.SolrResourceNotFoundException: Can't find resource 'countries.txt' in classpath or '/var/solr/data/search'
2026-02-27T11:37:05.156538206Z 	at org.apache.solr.spelling.suggest.FileDictionaryFactory.create(FileDictionaryFactory.java:58) ~[?:?]
```
Did I forget something? I used the solr image created by the pipelines in this PR.

EikLoe commented:
Hi @kshepherd ,
yes, If I copy the file, it works, thank you! I like this approach and it works for me! 👍🏾

kshepherd committed:
Public reload config method for MetadataExposureService

This is useful in test scope, and for the ability to reload
an urgent change in live data

kshepherd committed:
Include --suggest (search suggest handler dict) rebuild in IndexClient

kshepherd committed:
Methods to allow admins to rebuild one or more solr suggest dictionaries

kshepherd committed:
Do not index vals to _suggest if item or md is restricted in any way

Item must be in archive, discoverable, not withdrawn, public READ
Metadata must not be isHidden according to exposure service

kshepherd committed:
Tests for metadata privacy, item privacy, flat file lookup

milanmajchrak commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceIndexSuggestFieldPlugin.java:
```diff
@@ -0,0 +1,107 @@
+
+    @Override
+    public void additionalIndex(Context context, IndexableObject idxObj, SolrInputDocument document) {
+        if (idxObj instanceof IndexableItem) {
```
Just my personal preference:
I think this would be easier to read with fewer nested if blocks, for example:
```
// Do not continue if the idxObj is not instance of IndexableItem
if (!(idxObj instanceof IndexableItem)) {
   return;
}
// Do not continue if the indexed object is not an Item
if (!(((IndexableItem) idxObj).getIndexedObject() instanceof Item item)) {
   return;
}
```

kshepherd commented on dspace-api/src/main/java/org/dspace/discovery/SolrSuggestService.java:
```diff
@@ -0,0 +1,117 @@
+                solrQuery.set("suggest.build", true);
+                solrQuery.setRequestHandler("/suggest");
+                QueryResponse response = solrClient.query(solrQuery);
+                log.error(response);
```
sorry, my own dev logging left in by accident, i've removed the log line altogether

kshepherd commented on dspace-api/src/main/java/org/dspace/discovery/SolrSuggestService.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    public void rebuildAllDictionaries() {
+        log.error("Rebuilding all dictionaries");
```
yep, fixed

kshepherd committed:
Remove dev logging from SubmissionFormConverter

kshepherd committed:
Remove dev logging from MetadataValidator

kshepherd committed:
Remove debug discovery logging from log4j-cli.xml

kshepherd commented on dspace/config/log4j2-cli.xml:
```diff
@@ -92,6 +92,8 @@
         <Logger name='org.dspace.utils'
                 level='ERROR'/>
+        <Logger name='org.dspace.discovery'
+                level='DEBUG'/>
```
left in after dev unintentionally, fixed

kshepherd commented:
> @kshepherd We tested it by configuring suggestions for both publishers and authors. The author suggestions were loaded from a `*.txt` file, and everything worked as expected.
> 
> We are wondering whether showing the weights is the most user-friendly approach here: <img alt="image" width="445" height="220" src="https://private-user-images.githubusercontent.com/90026355/561628136-03baf596-a55c-4f36-bcb8-c89f0d20e68f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzMyNjcyMTMsIm5iZiI6MTc3MzI2NjkxMywicGF0aCI6Ii85MDAyNjM1NS81NjE2MjgxMzYtMDNiYWY1OTYtYTU1Yy00ZjM2LWJjYjgtYzg5ZjBkMjBlNjhmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAzMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMzExVDIyMDgzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVmZTYyZjI4NjQ3YTQ4NDYwOGZhNTcyOGYzYTM0MjY1YzIzMjc1YjVkNzhlNWNjMDEyNjBmMzk2YTMwODgxNTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0WVKwDXZfkdueklYjFBEbsMicXWdn_WhHsTOskh5fng"> Would it be clearer without this information?

I personally agree, my first version used parantheses or badges to indicate weight (which is practically synonymous with "count" in our usage), but when we finalised dev, we were not sure if we should just leave it as is or reformat the display, or remove entirely - could we shift this discussion to the dspace-angular issue https://github.com/DSpace/dspace-angular/pull/4439 ?

milanmajchrak committed:
feat: make allowed-dictionaries config more restrictive

- Empty/null allowed-dictionaries list now denies all dictionaries (was: allow all)
- Wildcard '*' value allows all dictionaries
- Explicit list allows only named dictionaries
- Updated rebuildAllDictionaries() to handle wildcard and empty cases
- Updated default discovery.cfg to have no dictionaries allowed by default
- Added unit tests for SolrSuggestService.isAllowedDictionary()
- Added integration tests for empty list and wildcard behavior

Addresses review from DSpace#10855 (issuecomment-4039229476)

milanmajchrak committed:
Updated an author

milanmajchrak committed:
Remove wildcard (*) support from allowed-dictionaries

- Wildcard removed: Solr cannot rebuild all dictionaries without
  explicit names, so admins must list them explicitly.
- Removed manual trim(): Apache Commons already trims values.
- Removed null dictionary handling from createSuggestQuery()
  and sendBuildRequest() (dictionary is always required).
- Removed wildcard-related unit and integration tests.
- Updated discovery.cfg comments.

kshepherd committed:
Updating SubmissionFormFieldMatcher and ControllerIT for solr autocomplete tests
